docs/features/incremental-generators.cookbook.md

```diff
@@ -0,0 +1,810 @@
+partial class MyRecord
+{
+    public string Item1 { get; }
+    public int Item2 { get; }
+}
+```
+
+This attribute could also be used for [Participate in the IDE experience](#participate-in-the-ide-experience),
+when the full scope of that feature is fully designed. In that scenario,
+instead of the generator finding every type marked with the given attribute,
+the compiler would notify the generator of every type marked with the given
+attribute. For now we'll assume that the types are provided to us.
+
+The first task is to decide what we want our serialization to return. Let's say
+we do a simple JSON serialization that produces a string like the following
+
+```json
+{
+    "Item1": "abc",
+    "Item2": 11,
+}
+```
+
+For that we could add a `Serialize` method to our record type like the following:
+
+```C#
+public string Serialize()
+{
+    var sb = new StringBuilder();
+    sb.AppendLine("{");
+    int indent = 8;
+
+    // Body
+    addWithIndent($"\"Item1\": \"{this.Item1.ToString()}\",");
+    addWithIndent($"\"Item2\": {this.Item2.ToString()},");
+
+    sb.AppendLine("}");
+
+    return sb.ToString();
+
+    void addWithIndent(string s)
+    {
+        sb.Append(' ', indent);
+        sb.AppendLine(s);
+    }
+}
+```
+
+Obviously this is heavily simplified -- this example only handles the `string` and `int`
+types properly, adds a trailing comma to the json output and has no error recovery, but
+it should serve to demonstrate the kind of code a source generator could add to a compilation.
+
+Our next task is design a generator to generate the above code, since the
+above code is itself customized in the `// Body` section according to the
+actual properties in the class. In other words, we need to generate the code
+which will generate the JSON format. This is a generator-generator.
+
+Let's start with a basic template. We are adding a full source generator, so we'll need
+to generate a class with the same name as the input class, with a public method called
+`Serialize`, and a filler area where we write out the properties.
+
+```C#
+string template = @"
+using System.Text;
+partial class {0}
+{{
+    public string Serialize()
+    {{
+        var sb = new StringBuilder();
+        sb.AppendLine(""{{"");
+        int indent = 8;
+
+        // Body
+{1}
+
+        sb.AppendLine(""}}"");
+
+        return sb.ToString();
+
+        void addWithIndent(string s)
+        {{
+            sb.Append(' ', indent);
+            sb.AppendLine(s);
+        }}
+    }}
+}}";
+```
+
+Now that we know the general structure of the code, we need to examine the input
+type and find all the right info to fill in. This information is all available in
+a C# SyntaxTree in our example. Let's say we were given a `ClassDeclarationSyntax`
+that was confirmed to have a generation attribute attached. Then we could grab the
+name of the class and the name of it's properties as follows:
+
+```C#
+private static string Generate(ClassDeclarationSyntax c)
```

Comment:
This seems to hold onto syntax node in a discouraged way. Maybe we should remove this section from the incremental cookbook.

Reply:
Yeah. I tried to update the sample somewhat, but I agree with your assessment. I'll remove it.